### ISSUE TITLE
ReactorDispatcher: Making sure spec 1.3 is not violated and under race, signals are not lost upon concurrent ClosedChannelException 

### ISSUE THREAD

**Issue by anuchandy** [Azure/azure-sdk-for-java#27320]
### TL;DR

If there is a race in `ReactorDispatcher::signalWorkQueue` that results in concurrent throwing of `ClosedChannelException`, then reactive spec 1.3 may get violated and parts of SDK may not be able to recover.

#### Purpose of ReactorDispatcher::invoke

The `ReactorDispatcher::signalWorkQueue` is a private method that the public  API (scoped to implementation) `ReactorDispatcher::invoke` uses. Different parts of the Amqp library uses `ReactorDispatcher::invoke` API from multiple threads to schedule works (that touches ProtonJ thread-unsafe objects) to single-threaded ProtonJ message pump.

#### Concurrency safety on ReactorDispatcher::invoke

The data structure (ConcurrentLinkedQueue) used inside the `ReactorDispatcher::invoke` method is thread-safe and the pipe API for message pump notification is concurrent safe as well. Given `ReactorDispatcher::invoke` belongs to the hot path, these concurrent safeties help avoid marking the method `synchronized`.

#### Exceptions in ReactorDispatcher::invoke

The two expected exceptions from the guarding section of this method are - `RejectedExecutionException` REE (related to concurrency) and `IOException` IOE (if message pump pipe is closed). In places where the Amqp library invokes this method, the two exceptions are propagated to the Reactor pipeline (e.g., via Sink.emitError) for downstream to take appropriate action. 

A specific exception handled inside this method (i.e., an exception won't leave this method) is `ClosedChannelException` (CCE). CCE indicates that the thread executing the invoke method got interrupted, and as a result, JVM closes the message pump pipe. Therefore any resources that rely on the single-threaded ProtonJ message pump (attached to this pipe) are unusable. When this happens, a shutdown signal is emitted via `Sink.One` (which is by design can be completed once) on which the `ReactorConnection` object listen.

#### Issue related to signaling in ReactorDispatcher::invoke

In rare cases, MULTIPLE "interrupted" threads can concurrently bypass the guard check (which otherwise would have thrown `RejectedExecutionException` or `IOException`) and enter invoke method, then all those threads run into CCE. (Any threads coming in the future will get `IOException` since the pipe is closed.)

In the case of MULTIPLE "interrupted" threads, two things can happen -.

1. More than one of those "interrupted" threads can attempt to send shutdown signal to `Sink.One` "concurrently", which violates Reactor Spec. Rule 1.3. 
2. More than one of those "interrupted" threads can attempt to send shutdown signal  to `Sink.One` "serially"; by design, the Reactor library ignores the second or later emissions.

When `#1` happens, in some cases, we have seen it is going to Reactor global `onErrorDropped` hook, which becomes simply a log entry. Here is such an example log https://github.com/Azure/azure-sdk-for-java/issues/24762. The exact reason for going into global-hook needs to be identified, but the point is, we must not violate the spec 1.3.

> Following is the stack trace proving spec 1.3 violation due to concurrent `ClosedChannelException` (CCE).

```
reactor.core.Exceptions$ErrorCallbackNotImplemented: reactor.core.publisher.Sinks$EmissionException: Spec. Rule 1.3 - onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially.
Caused by: reactor.core.publisher.Sinks$EmissionException: Spec. Rule 1.3 - onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially.
	at reactor.core.publisher.InternalEmptySink.emitError(InternalEmptySink.java:76)
	at com.azure.core.amqp.implementation.ReactorDispatcher.signalWorkQueue(ReactorDispatcher.java:130)
	at com.azure.core.amqp.implementation.ReactorDispatcher.invoke(ReactorDispatcher.java:94)
	at com.azure.core.amqp.implementation.RequestResponseChannel.lambda$closeAsync$12(RequestResponseChannel.java:230)
	at reactor.core.publisher.MonoRunnable.subscribe(MonoRunnable.java:49)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4338)
        ....
```

When `#2` happens, it is as if those additional signals are lost; the `ReactorConnection` care only about the first Shutdown signal, there are no consumption of any future emitted values.

### Proposal:

The fix should make sure -
1. `invoke` do not violate spec 1.3,
2. the signal emission happens only once,
 3. to treat "interrupted" threads (those lost the battle of emission) just like any future threads receiving `RejectedExecutionException` under pipe close. Ideally, the downstream (callers of `invoke`) get to handle the side effect rather than waiting forever for a signal.
